lib/staccato/v4/event.rb

```diff
@@ -1,76 +1,145 @@
+          #
+          # Overrides the tracker timestamp_micros.
+          param :timestamp_micros
         end
+
+        events[klass.event_name] = klass
       end
 
-      events[event.event_name] = event
     end
 
-    def initialize(tracker, options = {})
-      self.tracker = tracker
-      self.options = Staccato::OptionSet.new(options)
+    module ClassMethods
+
+      attr_reader :recommended_params
+
+      def event_name
+        @event_name ||= name
+          .dup
+          .split('::')
+          .last
+          .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
+          .gsub(/([a-z\d])([A-Z])/, '\1_\2')
+          .tr('-', '_')
+          .downcase
+          .to_sym
+      end
+
+      def param(name)
+        # TODO: prevent assignment of restricted params
+        name = name.to_sym
+        recommended_params << name
+        def_delegator :@options, name
+      end
+
     end
 
-    # return the fields for this hit type
-    # @return [Array] the fields
-    def fields
-      self.class::FIELDS
+    def initialize(**options)
+      @options = Staccato::OptionSet.new(options)
+    end
+
+    def name
+      self.class.event_name
     end
 
     def as_json
-      {
-        name: name,
-        params: event_params
-      }
+      Staccato::V4.encode_params_json({name: name, params: params})
     end
 
-    private
+    def ==(other) other.class == self.class && other.as_json == as_json end
+    alias_method :eql?, :==
+
+    def hash; [self.class, as_json].hash end
+
+    # The event parameters to be sent, including custom parameters
+    # @return [Hash{Symbol=>Numeric,String,Array[Hash{Symbol=>Numeric,String}]}]
+    def params
+      Staccato::V4.encode_params_json options.to_h.compact
+    end
 
-    def event_params
-      fields.map { |field|
+    # The recommended parameter values for this event
+    # @return (see params)
+    def recommended_params
+      self.class.recommended_params.map {|field|
         [field, options[field]] unless options[field].nil?
       }.compact.to_h
     end
 
-    # include BooleanHelpers
-    # adds convert_booleans()
+    # The custom event parameters to be sent for this event
+    # @return (see params)
+    def custom_params
+      official = self.class.recommended_params
+      params.delete_if {|k, v| official.include? k }
+    end
+
+    # Adds and updates event parameters with the values in the params argument.
+    # @return self
+    def update(params)
+      params.to_hash.each_pair do |key, value|
+        options[key.to_sym] = value
+      end
+      self
+    end
+    alias_method :merge!, :update
+
+    # Adds parameters to the event, where the event parameter wasn't already set.
+    #
+    # Note: This behaves like `Hash#with_defaults!` from `ActiveSupport`.
+    #
+    # @return self
+    def with_defaults!(defaults)
+      defaults.to_hash.each_pair do |key, value|
+        key = key.to_sym
+        options[key] = value unless options.to_h.key?(key)
```

Comment:
Would `||=` work here?

If not, should `options` have a `has_key?`?

Reply:
I don't remember, but I might have copied this implementation from ActiveSupport.  They need to do it this way because they need to distinguish between `nil` values and no matching key.  We don't need to make that distinction, so `||=` would suffice.